### PR TITLE
TMA 807: Statistics collecting methods, trying to push data to splunk with STDERR

### PR DESCRIPTION
**Pull request by kurecka**

### CODE REVIEW

File: lib/gooddata/core/logging.rb [hunk 1]
```diff
@@ -60,7 +65,7 @@ def logging_on?
     #
     #     GoodData.logging_http_on
     #
-    def logging_http_on(level = DEFAULT_RESTLOG_LEVEL, output = DEFAULT_RESTLOG_OUTPUT, klass = DEFAULT_RESTLOGGER_CLASS)
+    def logging_http_on(level = DEFAULT_RESTLOG_LEVEL, output = DEFAULT_SPLUNKLOG_OUTPUT, klass = DEFAULT_SPLUNKLOGGER_CLASS)
```
Comment (yenkins) [inline lib/gooddata/core/logging.rb]: Similar code found in logging.rb:42, logging.rb:68, logging.rb:88 (mass = 48)

File: lib/gooddata/core/logging.rb [hunk 2]
```diff
@@ -80,6 +85,20 @@ def logging_http_on?
       !@rest_logger.instance_of?(NilLogger)
     end
 
+    def logging_splunk_on(level = DEFAULT_SPLUNKLOG_LEVEL, output = DEFAULT_SPLUNKLOG_OUTPUT, klass = DEFAULT_SPLUNKLOGGER_CLASS)
```
Comment (yenkins) [inline lib/gooddata/core/logging.rb]: Similar code found in logging.rb:42, logging.rb:68, logging.rb:88 (mass = 48)

File: lib/gooddata/rest/connection.rb [hunk 3]
```diff
@@ -382,53 +389,47 @@ def sst_token
...
+      end
+
+      def stats_log(brick_id)
+        log_res = ""
+        stats[:calls].each do |row|
+          log_string = "[" + row[:time_stamp] + "] "
+          row.each do |pair|
+            next if pair[0] == :time_stamp
+            log_string += pair[0].to_s + "="
+            log_string += "\"" if pair[0] == :endpoint || pair[0] == :time_stamp
```
Comment (yenkins) [inline lib/gooddata/rest/connection.rb]: Identical code found in connection.rb:425, connection.rb:427 (mass = 64)

Comment (yenkins) [inline lib/gooddata/rest/connection.rb]: Calls 'pair[0] == :endpoint' 2 times (DuplicateMethodCall)

File: lib/gooddata/rest/connection.rb [hunk 4]
```diff
@@ -382,53 +389,47 @@ def sst_token
...
+            log_string += pair[1].to_s
+            log_string += "\"" if pair[0] == :endpoint || pair[0] == :time_stamp
```
Comment (yenkins) [inline lib/gooddata/rest/connection.rb]: Identical code found in connection.rb:425, connection.rb:427 (mass = 64)

File: spec/unit/rest/logging_spec.rb [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+    )
+    stub_request(:post, "#{@server}/gdc/account/login")
+      .to_return(:body => { :userLogin => { :profile => "/profile/123" } }.to_json, :headers => { 'Content-Type' => "application/json" })
+    stub_request(:get, "#{@server}/gdc/account/token")
+      .to_return(:body => {}.to_json, :headers => { 'Content-Type' => "application/json" })
+    stub_request(:get, "#{@server}/profile/123")
+      .to_return(:body => {}.to_json, :headers => { 'Content-Type' => "application/json" })
+    stub_request(:get, "#{@server}/poll_test")
+      .to_return(:body => { stuff: :aaa }.to_json, :headers => { 'Content-Type' => "application/json" })
+    stub_request(:get, "#{@server}/too_many_reqs")
```
Comment (yenkins) [inline spec/unit/rest/logging_spec.rb]: Similar code found in logging_spec.rb:49, logging_spec.rb:51, logging_spec.rb:53 (mass = 51)

File: spec/unit/rest/logging_spec.rb [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+      .to_return(:body => {}.to_json, :status => 429, :headers => { 'Content-Type' => "application/json" })
+    stub_request(:get, "#{@server}/out_of_service")
```
Comment (yenkins) [inline spec/unit/rest/logging_spec.rb]: Similar code found in logging_spec.rb:49, logging_spec.rb:51, logging_spec.rb:53 (mass = 51)

File: spec/unit/rest/logging_spec.rb [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+      .to_return(:body => {}.to_json, :status => 503, :headers => { 'Content-Type' => "application/json" })
+    stub_request(:get, "#{@server}/internal_error")
```
Comment (yenkins) [inline spec/unit/rest/logging_spec.rb]: Similar code found in logging_spec.rb:49, logging_spec.rb:51, logging_spec.rb:53 (mass = 51)

File: lib/gooddata/bricks/middleware/gooddata_middleware.rb [hunk 8]
```diff
@@ -102,7 +108,8 @@ def connect(server, verify_ssl, username, password, sst_token) # rubocop:disable
             puts "Connecting as #{username} to server #{server}"
             conn = GoodData.connect(username, password, server: server, verify_ssl: verify_ssl)
           end
-          conn.stats_on
+
+          conn.stats_on if collect_stats
```
Comment (yenkins) [inline lib/gooddata/bricks/middleware/gooddata_middleware.rb]: Is controlled by argument 'collect_stats' (ControlParameter)

File: lib/gooddata/bricks/middleware/gooddata_middleware.rb [hunk 9]
```diff
@@ -24,6 +24,10 @@ def call(params)
         # Set logger
         logger = params['GDC_LOGGER']
         GoodData.logger = logger
+        splunk_logger = params['GDC_SPLUNK_LOGGER']
```
Comment (yenkins) [inline lib/gooddata/bricks/middleware/gooddata_middleware.rb]: Refers to 'params' more than self (maybe move it to another class?) (FeatureEnvy)

File: lib/gooddata/bricks/middleware/gooddata_middleware.rb [hunk 10]
```diff
@@ -93,7 +99,7 @@ def call(params)
       end
 
       class << self
-        def connect(server, verify_ssl, username, password, sst_token) # rubocop:disable Metrics/ParameterLists
+        def connect(server, verify_ssl, username, password, sst_token, collect_stats) # rubocop:disable Metrics/ParameterLists
```
Comment (yenkins) [inline lib/gooddata/bricks/middleware/gooddata_middleware.rb]: Has 6 parameters (LongParameterList)

Comment (yenkins) [inline lib/gooddata/bricks/middleware/gooddata_middleware.rb]: Has approx 7 statements (TooManyStatements)

File: lib/gooddata/bricks/middleware/logger_middleware.rb [hunk 11]
```diff
@@ -14,14 +14,23 @@ class LoggerMiddleware < Bricks::Middleware
...
         if params['GDC_LOGGING_OFF']
           logger = NilLogger.new
         else
           logger = params[:GDC_LOGGER_FILE].nil? ? Logger.new(STDOUT) : Logger.new(params[:GDC_LOGGER_FILE])
           logger.level = params['GDC_LOG_LEVEL'] if params['GDC_LOG_LEVEL']
           logger.info('Pipeline starts')
+
+          original_formatter = Logger::Formatter.new
+          logger.formatter = proc { |severity, datetime, progname, msg, to_splunk=params['SPLUNK_LOG_ALL']|
+            GoodData.splunk_logger.formatter.call(severity, datetime, progname, msg.dump) if to_splunk.to_s == "true"
```
Comment (yenkins) [inline lib/gooddata/bricks/middleware/logger_middleware.rb]: Calls 'msg.dump' 2 times (DuplicateMethodCall)

File: lib/gooddata/bricks/middleware/logger_middleware.rb [hunk 12]
```diff
@@ -14,14 +14,23 @@ class LoggerMiddleware < Bricks::Middleware
...
+
         if params['GDC_LOGGING_OFF']
           logger = NilLogger.new
         else
           logger = params[:GDC_LOGGER_FILE].nil? ? Logger.new(STDOUT) : Logger.new(params[:GDC_LOGGER_FILE])
           logger.level = params['GDC_LOG_LEVEL'] if params['GDC_LOG_LEVEL']
           logger.info('Pipeline starts')
+
+          original_formatter = Logger::Formatter.new
+          logger.formatter = proc { |severity, datetime, progname, msg, to_splunk=params['SPLUNK_LOG_ALL']|
```
Comment (yenkins) [inline lib/gooddata/bricks/middleware/logger_middleware.rb]: Refers to 'params' more than self (maybe move it to another class?) (FeatureEnvy)

Comment (yenkins) [inline lib/gooddata/bricks/middleware/logger_middleware.rb]: Style/SpaceAroundEqualsInParameterDefault: Surrounding space missing in default value assignment.

File: lib/gooddata/core/logging.rb [hunk 13]
```diff
@@ -17,8 +18,12 @@ module GoodData
   DEFAULT_RESTLOG_OUTPUT = STDOUT
   DEFAULT_RESTLOGGER_CLASS = Logger
 
+  DEFAULT_SPLUNKLOG_LEVEL = Logger::WARN
+  DEFAULT_SPLUNKLOG_OUTPUT = STDERR
+  DEFAULT_SPLUNKLOGGER_CLASS = SplunkLogger
+
   class << self
-    attr_accessor :logger, :rest_logger
+    attr_accessor :logger, :rest_logger, :splunk_logger
```
Comment (yenkins) [inline lib/gooddata/core/logging.rb]: Is a writable attribute (Attribute)

File: lib/gooddata/core/splunk_logger.rb [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+      @formatter = Logger::Formatter.new
+
+      @logs = ""
+    end
+
+    def clear_logs
+      @logs = ""
+    end
+
+    def log(log, time, buffered=true)
```
Comment (yenkins) [inline lib/gooddata/core/splunk_logger.rb]: Has boolean parameter 'buffered' (BooleanParameter)

Comment (yenkins) [inline lib/gooddata/core/splunk_logger.rb]: Style/SpaceAroundEqualsInParameterDefault: Surrounding space missing in default value assignment.

File: lib/gooddata/core/splunk_logger.rb [hunk 15]
```diff
@@ -0,0 +1,37 @@
...
+      if buffered == true
```
Comment (yenkins) [inline lib/gooddata/core/splunk_logger.rb]: Is controlled by argument 'buffered' (ControlParameter)

File: lib/gooddata/helpers/splunk_helper.rb [hunk 16]
```diff
@@ -0,0 +1,51 @@
...
+require 'nokogiri'
+
+
+require_relative 'global_helpers'
+
+module GoodData
+  module Helpers
+    module SplunkHelper
+      class << self
+        @@host = 'https://localhost:8089'
```
Comment (yenkins) [inline lib/gooddata/helpers/splunk_helper.rb]: Declares the class variable '@@host' (ClassVariable)

Comment (yenkins) [inline lib/gooddata/helpers/splunk_helper.rb]: Style/ClassVars: Replace class var @@host with a class instance var.

File: lib/gooddata/helpers/splunk_helper.rb [hunk 17]
```diff
@@ -0,0 +1,51 @@
...
+        @@paths = {
```
Comment (yenkins) [inline lib/gooddata/helpers/splunk_helper.rb]: Declares the class variable '@@paths' (ClassVariable)

Comment (yenkins) [inline lib/gooddata/helpers/splunk_helper.rb]: Style/ClassVars: Replace class var @@paths with a class instance var.

File: lib/gooddata/helpers/splunk_helper.rb [hunk 18]
```diff
@@ -0,0 +1,51 @@
+require 'rest-client'
+require 'nokogiri'
+
+
+require_relative 'global_helpers'
+
+module GoodData
+  module Helpers
+    module SplunkHelper
```
Comment (yenkins) [inline lib/gooddata/helpers/splunk_helper.rb]: Has no descriptive comment (IrresponsibleModule)

File: lib/gooddata/helpers/splunk_helper.rb [hunk 19]
```diff
@@ -0,0 +1,51 @@
...
+      class << self
+        @@host = 'https://localhost:8089'
+        @@paths = {
+          :login => '/services/auth/login?cookie=1',
+          :insert => '/services/receivers/simple'
+        }
+
+        def post(url_token, data, session_key = '', query = '')
```
Comment (yenkins) [inline lib/gooddata/helpers/splunk_helper.rb]: Has 4 parameters (LongParameterList)

File: lib/gooddata/helpers/splunk_helper.rb [hunk 20]
```diff
@@ -0,0 +1,51 @@
...
+        end
+
+        def get(url_token, session_key = '', query = '')
+          return RestClient::Request.execute( :method => 'get', :url => @@host+@@paths[url_token]+"?"+query, :verify_ssl => false, :headers => {
+            "Authorization" => "Splunk "+session_key
+          })
+        end
+
+
+        def send_logs(logs)
```
Comment (yenkins) [inline lib/gooddata/helpers/splunk_helper.rb]: Has approx 7 statements (TooManyStatements)

File: lib/gooddata/helpers/splunk_helper.rb [hunk 21]
```diff
@@ -0,0 +1,51 @@
...
+            'password'=>'1234'
+          }
+
+          puts "SENDING TO SPLUNK:"
+          puts logs
+
+          response = post(:login, login_payload)
+
+          xml = Nokogiri::XML(response.body)
+          sessionKey = xml.at_xpath("//sessionKey").content
```
Comment (yenkins) [inline lib/gooddata/helpers/splunk_helper.rb]: Has the variable name 'sessionkey' (UncommunicativeVariableName)

Comment (yenkins) [inline lib/gooddata/helpers/splunk_helper.rb]: Style/VariableName: Use snake_case for variable names.

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-opensource-repository-check/256/ : SUCCESS in 11s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/1102/ : FAILURE in 4m 14s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/1371/ : FAILURE in 3m 39s

File: lib/gooddata/rest/connection.rb [hunk 22]
```diff
@@ -382,53 +390,47 @@ def sst_token
...
+      end
+
+      def stats_log(brick_id)
+        log_res = ""
+        stats[:calls].each do |row|
+          log_string = "[" + row[:time_stamp] + "] "
+          row.each do |pair|
+            next if pair[0] == :time_stamp
+            log_string += pair[0].to_s + "="
+            log_string += "\"" if pair[0] == :endpoint || pair[0] == :time_stamp
```
Comment (yenkins) [inline lib/gooddata/rest/connection.rb]: Identical code found in connection.rb:426, connection.rb:428 (mass = 64)

File: lib/gooddata/rest/connection.rb [hunk 23]
```diff
@@ -382,53 +390,47 @@ def sst_token
...
+            log_string += pair[1].to_s
+            log_string += "\"" if pair[0] == :endpoint || pair[0] == :time_stamp
```
Comment (yenkins) [inline lib/gooddata/rest/connection.rb]: Identical code found in connection.rb:426, connection.rb:428 (mass = 64)

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-opensource-repository-check/257/ : SUCCESS in 8s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/1103/ : FAILURE in 1m 16s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/1372/ : FAILURE in 3m 28s

File: lib/gooddata/core/logging.rb [hunk 24]
```diff
@@ -60,7 +66,7 @@ def logging_on?
     #
     #     GoodData.logging_http_on
     #
-    def logging_http_on(level = DEFAULT_RESTLOG_LEVEL, output = DEFAULT_RESTLOG_OUTPUT, klass = DEFAULT_RESTLOGGER_CLASS)
+    def logging_http_on(level = DEFAULT_RESTLOG_LEVEL, output = DEFAULT_SPLUNKLOG_OUTPUT, klass = DEFAULT_SPLUNKLOGGER_CLASS)
```
Comment (yenkins) [inline lib/gooddata/core/logging.rb]: Similar code found in logging.rb:43, logging.rb:69 (mass = 32)

File: lib/gooddata/bricks/middleware/logger_middleware.rb [hunk 25]
```diff
@@ -14,14 +14,28 @@ class LoggerMiddleware < Bricks::Middleware
...
+            GoodData.splunk_logger.formatter.call(severity, datetime, progname, msg.dump) if to_splunk.to_s == "true"
+            original_formatter.call(severity, datetime, progname, msg.dump)
+          }
         end
         params['GDC_LOGGER'] = logger
+
+        if params['COLLECT_STATS'] && params['COLLECT_STATS'].to_b
```
Comment (yenkins) [inline lib/gooddata/bricks/middleware/logger_middleware.rb]: Calls 'params['collect_stats']' 2 times (DuplicateMethodCall)

File: lib/gooddata/core/logging.rb [hunk 26]
```diff
@@ -17,8 +18,13 @@ module GoodData
...
   DEFAULT_RESTLOGGER_CLASS = Logger
 
+  DEFAULT_SPLUNKLOG_LEVEL = Logger::INFO
+  DEFAULT_SPLUNKLOG_OUTPUT = STDERR
+  DEFAULT_SPLUNKLOGGER_CLASS = SplunkLogger
+  DEFAULT_SPLUNKLOG_MODE = SplunkLogger::FILE_MODE | SplunkLogger::BUFFERED
+
   class << self
-    attr_accessor :logger, :rest_logger
+    attr_accessor :logger, :rest_logger, :splunk_logger
```
Comment (yenkins) [inline lib/gooddata/core/logging.rb]: Is a writable attribute (Attribute)

File: lib/gooddata/core/logging.rb [hunk 27]
```diff
@@ -80,6 +86,21 @@ def logging_http_on?
       !@rest_logger.instance_of?(NilLogger)
     end
 
+    def logging_splunk_on(level = DEFAULT_SPLUNKLOG_LEVEL, output = DEFAULT_SPLUNKLOG_OUTPUT, klass = DEFAULT_SPLUNKLOGGER_CLASS, mode = DEFAULT_SPLUNKLOG_MODE)
```
Comment (yenkins) [inline lib/gooddata/core/logging.rb]: Has 4 parameters (LongParameterList)

File: lib/gooddata/core/splunk_logger.rb [hunk 28]
```diff
@@ -0,0 +1,82 @@
...
+    def set_context(key, val)
+      @context[key] = val
+    end
+
+    def clear_logs
+      @logs = ""
+    end
+
+    def extract_mode(mode)
+      return mode == (mode & @mode)
```
Comment (yenkins) [inline lib/gooddata/core/splunk_logger.rb]: Refers to 'mode' more than self (maybe move it to another class?) (FeatureEnvy)

Comment (yenkins) [inline lib/gooddata/core/splunk_logger.rb]: Style/RedundantReturn: Redundant `return` detected.

File: lib/gooddata/core/splunk_logger.rb [hunk 29]
```diff
@@ -0,0 +1,82 @@
...
+    end
+
+    def add(severity, message = nil, progname = nil)
+      context = SplunkLogger.hash_to_string(@context)
+      if message.class == Hash
+        message = SplunkLogger.hash_to_string(message) + context
+      elsif message.nil?
```
Comment (yenkins) [inline lib/gooddata/core/splunk_logger.rb]: Performs a nil-check (NilCheck)

File: lib/gooddata/core/splunk_logger.rb [hunk 30]
```diff
@@ -0,0 +1,82 @@
...
+
+    def clear_logs
+      @logs = ""
+    end
+
+    def extract_mode(mode)
+      return mode == (mode & @mode)
+    end
+
+    def add(severity, message = nil, progname = nil)
```
Comment (yenkins) [inline lib/gooddata/core/splunk_logger.rb]: Has approx 9 statements (TooManyStatements)

File: lib/gooddata/core/splunk_logger.rb [hunk 31]
```diff
@@ -0,0 +1,82 @@
...
+  class SplunkLogger < Logger
+    BUFFERED = 1 << 0
+    API_MODE = 1 << 1
+    FILE_MODE = 1 << 2
+
+    BRICK_CONTEXT = :brick
+    ACTION_CONTEXT = :action
+
+    class << self
+      def hash_to_string(hash)
```
Comment (yenkins) [inline lib/gooddata/core/splunk_logger.rb]: Has approx 6 statements (TooManyStatements)

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-opensource-repository-check/265/ : SUCCESS in 10s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/1115/ : FAILURE in 2m 16s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/1384/ : FAILURE in 3m 10s

**Comment by gdgate** [comment]
Merge Failed.

This change or one of its cross-repo dependencies was unable to be automatically merged with the current state of its repository. Please rebase the change and upload a new patchset.

**Comment by gdgate** [comment]
Merge Failed.

This change or one of its cross-repo dependencies was unable to be automatically merged with the current state of its repository. Please rebase the change and upload a new patchset.

**Pull request closed without merging**